File: plugins/agent-skills/skills/claude-delegate/scripts/claude_delegate/job_state.py
from __future__ import annotations

from pathlib import Path

from .common import read_json, utc_now, write_json


def build_job_record(paths: dict[str, Path], request: dict) -> dict:
    workflow_roles = request.get("workflow_roles", [request.get("assistant_role")])
    return {
        "assistant_role": request.get("assistant_role"),
        "cancel_requested_at": None,
        "completed_roles": [],
        "created_at": request["created_at"],
        "current_role": None if len(workflow_roles) > 1 else request.get("assistant_role"),
        "current_step_job_path": None,
        "delegate_pid": None,
        "event_count": 0,
        "events_path": str(paths["events"]),
        "finished_at": None,
        "job_id": paths["artifacts_dir"].name,
        "job_path": str(paths["artifacts_dir"]),
        "last_error": None,
        "last_event_at": None,
        "last_event_type": None,
        "lock_path": str(paths["lock"]),
        "normalized_path": str(paths["normalized"]),
        "artifact_lifecycle": None,
        "pause_requested_at": None,
        "pid": None,
        "request_path": str(paths["request"]),
        "routing": request.get("routing"),
        "started_at": None,
        "state": "submitted",
        "step_job_paths": [],
        "stderr_path": str(paths["stderr"]),
        "stdout_path": str(paths["stdout"]),
        "task_type": request.get("task_type") or request.get("task_packet", {}).get("task_type"),
        "termination_intent": None,
        "updated_at": request["created_at"],
        "workflow_roles": workflow_roles,
        "workflow_step": 0,
        "workflow_total_steps": len(workflow_roles),
    }


def initialize_job(paths: dict[str, Path], request: dict, *, ledger_path: str | None = None) -> dict:
    record = build_job_record(paths, request)
    if ledger_path is not None:
        record["ledger_path"] = ledger_path
    write_json(paths["job"], record)
    return record


def update_job(paths: dict[str, Path], **changes: object) -> dict:
    job = read_json(paths["job"]) or {}
    job.update(changes)
    job["updated_at"] = utc_now()
    write_json(paths["job"], job)
    return job


def finalize_job_state(envelope: dict) -> str:
    if envelope.get("ok"):
        return "finished"
    if envelope.get("error_type") == "paused":
        return "paused"
    if envelope.get("error_type") == "cancelled":
        return "cancelled"
    return "failed"

File: plugins/agent-skills/skills/claude-delegate/src/lib/job-state.ts
import { readJson, writeJson, utcNow } from "./common.js";
import type { ArtifactPaths } from "./common.js";

export function buildJobRecord(
  paths: ArtifactPaths,
  request: Record<string, unknown>,
): Record<string, unknown> {
  const workflowRoles = (request.workflow_roles as string[]) ?? [
    request.assistant_role as string,
  ];
  return {
    assistant_role: request.assistant_role ?? null,
    cancel_requested_at: null,
    completed_roles: [],
    created_at: request.created_at,
    current_role:
      workflowRoles.length > 1
        ? null
        : (request.assistant_role as string) ?? null,
    current_step_job_path: null,
    delegate_pid: null,
    event_count: 0,
    events_path: paths.events,
    finished_at: null,
    job_id: paths.artifactsDir.split("/").pop()!,
    job_path: paths.artifactsDir,
    last_error: null,
    last_event_at: null,
    last_event_type: null,
    lock_path: paths.lock,
    normalized_path: paths.normalized,
    artifact_lifecycle: null,
    pause_requested_at: null,
    pid: null,
    request_path: paths.request,
    routing: request.routing ?? null,
    runtime: request.runtime ?? null,
    started_at: null,
    state: "submitted",
    step_job_paths: [],
    stderr_path: paths.stderr,
    stdout_path: paths.stdout,
    task_type:
      (request.task_type as string) ??
      ((request.task_packet as Record<string, unknown>) ?? {}).task_type ??
      null,
    termination_intent: null,
    updated_at: request.created_at,
    workflow_roles: workflowRoles,
    workflow_step: 0,
    workflow_total_steps: workflowRoles.length,
  };
}

export function initializeJob(
  paths: ArtifactPaths,
  request: Record<string, unknown>,
  opts?: { ledgerPath?: string },
): Record<string, unknown> {
  const record = buildJobRecord(paths, request);
  if (opts?.ledgerPath !== undefined) {
    record.ledger_path = opts.ledgerPath;
  }
  writeJson(paths.job, record);
  return record;
}

export function updateJob(
  paths: ArtifactPaths,
  changes: Record<string, unknown>,
): Record<string, unknown> {
  const job = readJson(paths.job) ?? {};
  Object.assign(job, changes);
  job.updated_at = utcNow();
  writeJson(paths.job, job);
  return job;
}

export function finalizeJobState(envelope: Record<string, unknown>): string {
  if (envelope.ok) return "finished";
  if (envelope.error_type === "paused") return "paused";
  if (envelope.error_type === "cancelled") return "cancelled";
  return "failed";
}
